# Data Analytics — Exercise 7.2: Pandas Essentials
## Sharleen Guerrero | Date: 5/19/2026

This notebook uses the Pandas library to load, inspect, filter, and summarise employee sales data for RetailHub Inc. It covers DataFrame creation, Boolean filtering, calculated columns, apply() with a lambda, and GroupBy aggregation.

## 1. Import Pandas and Create the DataFrame
We import Pandas and build a DataFrame from a dictionary of employee records.

In [5]:
# A DataFrame is like a table with rows, columns, and indexes.
# It's like Pandas version of NumPy's arrays, but with labeled columns and mixed data types

import pandas as pd

# Requirement 1 — import Pandas with standard alias
import pandas as pd

# Requirement 2 — employee sales data (8 records, 4 columns)
# Step 1 is create a dictionary with keys (columns) and values (rows)
data = {
    'name'   : ['Tommy', 'MollyMae', 'Maura', 'India', 'Amber', 'Greg', 'Yewande', 'Ovie'],
    'city'   : ['London', 'Manchester', 'London', 'Leeds', 'Birmingham', 'Manchester', 'Leeds', 'London'],
    'sales'  : [4200.0, 3100.0, 5050.0, 2800.0, 1950.0, 3750.0, 4600.0, 3300.0],
    'months' : [12, 6, 24, 3, 18, 9, 15, 7]
}

# Step 2 is turning the dictionary into a Dataframe
# df is a variable name to store the table, .pd is pandas, and DataFrame() is the function 
# that turns it into a Pandas table. (data)
df = pd.DataFrame(data)

print('── Employee DataFrame ──────────────────────')
print(df)                       # prints whole table
print(f'\nShape : {df.shape}')  # df.shape shows the rows and columns 
print(f'dtypes :\n{df.dtypes}') # df.dtpes shows data types of each column
print('\ndf.info():')           # df.info() is a summary
df.info()

# non-null means the cell has a value
# here, df.info() is saying "data clean, no missing values)

── Employee DataFrame ──────────────────────
       name        city   sales  months
0     Tommy      London  4200.0      12
1  MollyMae  Manchester  3100.0       6
2     Maura      London  5050.0      24
3     India       Leeds  2800.0       3
4     Amber  Birmingham  1950.0      18
5      Greg  Manchester  3750.0       9
6   Yewande       Leeds  4600.0      15
7      Ovie      London  3300.0       7

Shape : (8, 4)
dtypes :
name          str
city          str
sales     float64
months      int64
dtype: object

df.info():
<class 'pandas.DataFrame'>
RangeIndex: 8 entries, 0 to 7
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   name    8 non-null      str    
 1   city    8 non-null      str    
 2   sales   8 non-null      float64
 3   months  8 non-null      int64  
dtypes: float64(1), int64(1), str(2)
memory usage: 388.0 bytes


## 2. Inspect and Describe the Data
We use Pandas methods to preview the data and compute summary statistics.

In [9]:
print("First 3 rows:")
print(df.head(3))
# Head() means to show the top 3 rows, if rows aren't specified it will default to 5 rows

print("\nLast 3 rows:")
print(df.tail(3))
# This will show the bottom 3 rows

# 0, 1, 2, 5, 6, 7 are automatic index numbers pandas assigns to every row. 0 Tommy is row 0 (first row)

print("\nSummary Statistics:")
print(df.describe())
# describe() is a pandas method, it will you summary statistics for all number columns automatically

print(f"\nMean sales: £{df['sales'].mean():,.2f}")
print(f"Max sales: £{df['sales'].max():,.2f}")

# df["sales"] looks into the sales column from df DataFrame
# .mean() calculates average 
# .max() gets the highest value
# £ identifes the currency as pounds
# format to 2 decimal places with comma separator

First 3 rows:
       name        city   sales  months
0     Tommy      London  4200.0      12
1  MollyMae  Manchester  3100.0       6
2     Maura      London  5050.0      24

Last 3 rows:
      name        city   sales  months
5     Greg  Manchester  3750.0       9
6  Yewande       Leeds  4600.0      15
7     Ovie      London  3300.0       7

Summary Statistics:
             sales     months
count     8.000000   8.000000
mean   3593.750000  11.750000
std    1013.282326   6.964194
min    1950.000000   3.000000
25%    3025.000000   6.750000
50%    3525.000000  10.500000
75%    4300.000000  15.750000
max    5050.000000  24.000000

Mean sales: £3,593.75
Max sales: £5,050.00


## 3. Select, Filter, and Sort
Pandas makes it easy to slice columns, filter rows, and sort results.

In [11]:
# Selects two columns from df DataFrame
# single bracket is for one column
# Double bracket is for more than one columns
# Here we are just printint whats in those columns and index position
print('Name & Sales columns:')
print(df[['name', 'sales']])

# Boolean filter — sales above £3500
high_sales = df[df['sales'] > 3500.0]
print('\nSales > £3500:')
print(high_sales)
# df['sales'] > 3500 (inner brackets) checks every sales value in the DataFrame
# and returns True or False for each row
# df[...] (outer brackets) then uses those True/False results
# to filter the DataFrame, only keeping rows where the condition is True
# the result is stored in high_sales

# sort_values() is a Pandas method that sorts the DataFrame by a specified column
# ascending=False sorts from highest to lowest (descending)
# Pandas defaults to ascending=True (lowest to highest) so we turn it off
# reset_index(drop=True) resets the index back to 0-7 after sorting, but it's not necessary
# however keeping the original index can be useful to know where each row came from
# drop=True prevents the old index from being added as an extra column
df_sorted = df.sort_values('sales', ascending=False).reset_index(drop=True)
print('\nSorted by sales (highest first):')
print(df_sorted)

# Top seller — first row after sort
print('\nTop seller record:')
print(df_sorted.iloc[0])
# # iloc means integer location
# iloc[0] gets the row at position 0
# after sorting, position 0 is the top seller 
# # since we already sorted by sales descending, the highest seller is at position 0
# so iloc[0] automatically gets the top seller (Maura with £5050)
# if we hadn't sorted first, iloc[0] would have returned Tommy (original first row)

Name & Sales columns:
       name   sales
0     Tommy  4200.0
1  MollyMae  3100.0
2     Maura  5050.0
3     India  2800.0
4     Amber  1950.0
5      Greg  3750.0
6   Yewande  4600.0
7      Ovie  3300.0

Sales > £3500:
      name        city   sales  months
0    Tommy      London  4200.0      12
2    Maura      London  5050.0      24
5     Greg  Manchester  3750.0       9
6  Yewande       Leeds  4600.0      15

Sorted by sales (highest first):
       name        city   sales  months
0     Maura      London  5050.0      24
1   Yewande       Leeds  4600.0      15
2     Tommy      London  4200.0      12
3      Greg  Manchester  3750.0       9
4      Ovie      London  3300.0       7
5  MollyMae  Manchester  3100.0       6
6     India       Leeds  2800.0       3
7     Amber  Birmingham  1950.0      18

Top seller record:
name       Maura
city      London
sales     5050.0
months        24
Name: 0, dtype: object


## 4. Calculated Column and Conditional Rating
We add commission (8% of sales) and a performance rating

In [14]:
# Add commission column — 8% of sales
# Creates a brand new column named 'commission' by wrapping it in df[]
# multiplies every value in the existing 'sales' column by 0.08 (8%)
# and stores the results in the new 'commission' column
df['commission'] = df['sales'] * 0.08

# Add rating using apply() with a lambda
# .apply() will go through every row and apply the function on every value in the column
# lambda x: will take one sales value at a time and see if the first condition is met
# If it failed, it will test for the second condition and finally if both failed it will print out 'Developing'
df['rating'] = df['sales'].apply(
    lambda x: 'Star' if x >= 4000
    else ('Good' if x >= 2500 else 'Developing')
)

print('Updated DataFrame:')
print(df)
# This prints the two new columns that were made, commission and rating

# loop through every row in the DataFrame one at a time
# ignore the index number (_) and only use the row data
# for each row, print the employee name, their commission formatted to 2dp, and their rating
for _, row in df.iterrows():
    print(f'{row["name"]:<8} | £{row["commission"]:>8,.2f} | {row["rating"]}')

Updated DataFrame:
       name        city   sales  months  commission      rating
0     Tommy      London  4200.0      12       336.0        Star
1  MollyMae  Manchester  3100.0       6       248.0        Good
2     Maura      London  5050.0      24       404.0        Star
3     India       Leeds  2800.0       3       224.0        Good
4     Amber  Birmingham  1950.0      18       156.0  Developing
5      Greg  Manchester  3750.0       9       300.0        Good
6   Yewande       Leeds  4600.0      15       368.0        Star
7      Ovie      London  3300.0       7       264.0        Good
Tommy    | £  336.00 | Star
MollyMae | £  248.00 | Good
Maura    | £  404.00 | Star
India    | £  224.00 | Good
Amber    | £  156.00 | Developing
Greg     | £  300.00 | Good
Yewande  | £  368.00 | Star
Ovie     | £  264.00 | Good


## 5. GroupBy — Sales by City
GroupBy lets us aggregate data by a category, revealing city-level performance.

In [15]:
# groupby() groups all rows that share the same city together
# then .sum() adds up all the sales values for each city group
city_totals = df.groupby('city')['sales'].sum()
print('── Total Sales by City ──────────────────')
print(city_totals)

# .mean() gets the average sales for each city group
city_means = df.groupby('city')['sales'].mean()
print('\n── Mean Sales by City ───────────────────')
print(city_means)

# .agg() lets you compute multiple statistics in one call
# instead of calling sum, mean, count separately
city_agg = df.groupby('city').agg({'sales': ['sum', 'mean', 'count']})
print('\n── Full City Summary ────────────────────')
print(city_agg)

# idxmax() returns the city name with the highest total sales
top_city = city_totals.idxmax()
print(f'\nTop city: {top_city} Total: £{city_totals[top_city]:,.2f}')

── Total Sales by City ──────────────────
city
Birmingham     1950.0
Leeds          7400.0
London        12550.0
Manchester     6850.0
Name: sales, dtype: float64

── Mean Sales by City ───────────────────
city
Birmingham    1950.000000
Leeds         3700.000000
London        4183.333333
Manchester    3425.000000
Name: sales, dtype: float64

── Full City Summary ────────────────────
              sales                   
                sum         mean count
city                                  
Birmingham   1950.0  1950.000000     1
Leeds        7400.0  3700.000000     2
London      12550.0  4183.333333     3
Manchester   6850.0  3425.000000     2

Top city: London Total: £12,550.00


## Summary
- **London** is the top-performing city with a combined total of **£12,550** from 3 employees.
- **3 of 8 employees** achieved Star rating (Tommy, Maura, Yewande); only Amber (Birmingham) is Developing.
- The team's average monthly sales of **£3,593.75** and top performer Maura (£5,050) suggest London should be prioritised for investment in the next quarter.